File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        
    int a = 0;

    for (int i = 0; i < grid.length;  ++i)
        
      for (int j = 0; j < grid[0].length; ++j)
          
        if (grid[i][j] == '1') {
            
          dfs(grid, i, j);
          ++a;
            
        }

        
    return a;
        
  }

  private void dfs(char[][] grid, int i, int j) {
      
    if (i < 0 || i == grid.length || j < 0 || j == grid[0].length)
      return;
      
    if (grid[i][j] != '1')
      return;

    grid[i][j] = '2'; // mark '2' as visited
      
    dfs(grid, i + 1, j);
    dfs(grid, i - 1, j);
    dfs(grid, i, j + 1);
    dfs(grid, i, j - 1);
  }
}
 

File: GO-Leet/200. Number of Islands.go
package main

func numIslands(grid [][]byte) int {
	res := 0

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == '1' {
				dfs(grid, i, j)
				res++
			}
		}
	}

	return res
}

func dfs(grid [][]byte, x int, y int) {
	if x < 0 || y < 0 || x >= len(grid) || y >= len(grid[0]) || grid[x][y] == '0' {
		return
	}

	grid[x][y] = '0'

	dfs(grid, x+1, y)
	dfs(grid, x, y+1)
	dfs(grid, x-1, y)
	dfs(grid, x, y-1)

	return
}
